### PR TITLE
Cos 263 Improve test coverage(change, database, precondition packages)

### PR DESCRIPTION
**Pull request by tigrankhojoyan**

### CODE REVIEW

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/CreateBucketChangeTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+class CreateBucketChangeTest {
+
+    private ChangeLogProvider changeLogProvider;
+
+    @BeforeEach
+    void setUp() {
+        CouchbaseLiquibaseDatabase db = mock(CouchbaseLiquibaseDatabase.class);
+        changeLogProvider = new TestChangeLogProvider(db);
+    }
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/CreateBucketChangeTest.java]: I think we can mock 1 time and create provider 1 time. Using @Mock and @InjectMocks annotations. We have such examples in code you can look at them.

And the same for such code below

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/ExecuteQueryChangeTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import static common.constants.ChangeLogSampleFilePaths.EXECUTE_QUERY_CHANGE_TEST_XML;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.internal.util.collections.Iterables.firstOf;
+
+class ExecuteQueryChangeTest {
+
+    private ChangeLogProvider changeLogProvider;
+    private final String query = "DELETE FROM `testBucket`.testScope.testCollection WHERE META().id = $id";
+    List<Param> paramList = Lists.newArrayList(new Param("id", "abcd"));
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/ExecuteQueryChangeTest.java]: private?

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/InsertDocumentsChangeTest.java [hunk 3]
```diff
@@ -51,5 +62,56 @@ void Should_contains_specific_documents() {
...
+
+    private InsertDocumentsChange createInsertDocumentChange(String bucketName, String scopeName, String collectionName,
+                                                             List<Document> documents, File file) {
+        InsertDocumentsChange change = new InsertDocumentsChange();
+        change.setBucketName(bucketName);
+        change.setScopeName(scopeName);
+        change.setCollectionName(collectionName);
+        change.setDocuments(documents);
+        change.setFile(file);
+        return change;
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/InsertDocumentsChangeTest.java]: Maybe better to use Builder?

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/RemoveDocumentsChangeTest.java [hunk 4]
```diff
@@ -78,5 +85,40 @@ void Should_contain_where_clause() {
...
+    }
+
+    private RemoveDocumentsChange createRemoveDocumentChange() {
+        RemoveDocumentsChange change = new RemoveDocumentsChange();
+        change.setBucketName(TEST_BUCKET);
+        change.setScopeName(TEST_SCOPE);
+        change.setCollectionName(TEST_COLLECTION);
+        change.setIds(Sets.newHashSet(new Id("id1"), new Id("id2")));
+
+        return change;
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/RemoveDocumentsChangeTest.java]: Builder?

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/UpsertDocumentsChangeTest.java [hunk 5]
```diff
@@ -41,4 +53,57 @@ void Should_contains_specific_document() {
...
+    private UpsertDocumentsChange createUpsertDocumentChange(String bucketName, String scopeName, String collectionName,
+                                                             List<Document> documents, File file) {
+        UpsertDocumentsChange change = new UpsertDocumentsChange();
+        change.setBucketName(bucketName);
+        change.setScopeName(scopeName);
+        change.setCollectionName(collectionName);
+        change.setDocuments(documents);
+        change.setFile(file);
+        return change;
+    }
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/UpsertDocumentsChangeTest.java]: Builder? And the same for the same code below

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/database/CouchbaseConnectionTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import static org.mockito.Mockito.when;
+
+class CouchbaseConnectionTest {
+
+    private static final String DB_URL = "couchbase://127.0.0.1";
+    private final CouchbaseConnection connection = spy(CouchbaseConnection.class);
+    private final Database database = mock(Database.class);
+    private final Bucket bucket = mock(Bucket.class);
+    private final Cluster cluster = mock(Cluster.class);
+    Driver driver = mock(Driver.class);
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/database/CouchbaseConnectionTest.java]: private? and check everywhere below

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/database/CouchbaseConnectionTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        assertThat(connection.getConnectionString().scheme().name()).isEqualTo("COUCHBASE");
+        assertThat(connection.getCatalog()).isEqualTo(StringUtils.EMPTY);
+        assertThat(connection.getDatabaseProductName()).isEqualTo(COUCHBASE_PRODUCT_NAME);
+        assertThat(connection.getURL()).isEqualTo("127.0.0.1");
+        assertThat(connection.isClosed()).isEqualTo(false);
+        assertThat(connection.getAutoCommit()).isEqualTo(false);
+        assertThat(connection.getPriority()).isEqualTo(PRIORITY_DEFAULT + COUCHBASE_PRIORITY);
+        assertThat(connection.getDatabaseProductVersion()).isEqualTo("0");
+        assertThat(connection.getDatabaseMajorVersion()).isEqualTo(0);
+        assertThat(connection.getDatabaseMinorVersion()).isEqualTo(0);
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/database/CouchbaseConnectionTest.java:59]: Looks like massive assertion. CouchbaseConnection has @data annotation and it contains equals and hashcode. I think we can prepare CouchbaseConnection **expected** object and compare result with expected via 1 assertion. 
What do you think?

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/database/CouchbaseLiquibaseDatabaseTest.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    void Should_create_connection_correctly() {
+        CouchbaseLiquibaseDatabase couchbaseLiquibaseDatabase = new CouchbaseLiquibaseDatabase("user", "password", DB_URL);
+        couchbaseLiquibaseDatabase.getConnection();
+
+        assertThat(couchbaseLiquibaseDatabase.getDefaultDatabaseProductName()).isEqualTo(COUCHBASE_PRODUCT_NAME);
+        assertThat(couchbaseLiquibaseDatabase.getDefaultDriver(DB_URL)).isEqualTo(CouchbaseStubDriver.class.getName());
+        assertThat(couchbaseLiquibaseDatabase.supportsTablespaces()).isEqualTo(false);
+        assertThat(couchbaseLiquibaseDatabase.supportsInitiallyDeferrableColumns()).isEqualTo(false);
+        assertThat(couchbaseLiquibaseDatabase.getDefaultPort()).isEqualTo(DEFAULT_PORT);
+        assertThat(couchbaseLiquibaseDatabase.getShortName()).isEqualTo(COUCHBASE_PRODUCT_SHORT_NAME);
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/database/CouchbaseLiquibaseDatabaseTest.java:43]: Can we do it with 1 assertion like in the comment above?

Comment (tigrankhojoyan) [reply liquibase-couchbase/src/test/java/liquibase/ext/couchbase/database/CouchbaseLiquibaseDatabaseTest.java:43]: Good point, but in both cases we need to init an instance of connection(but the parameters are initialized via open method, no by setters). I think better to let it as is the adding setters for params, which are initialized after connection

Comment (DaimonCool) [reply liquibase-couchbase/src/test/java/liquibase/ext/couchbase/database/CouchbaseLiquibaseDatabaseTest.java:43]: Ohh no. I see now. You assert not database fields, but its methods. Sorry, my bad :)

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/CreatePrimaryQueryIndexChangeTest.java [hunk 9]
```diff
@@ -44,6 +38,7 @@ void Should_return_correct_confirmation_message() {
 
     @Test
     void Changelog_should_contain_correct_types_only() {
+        DatabaseChangeLog changeLog = changeLogProvider.load(CREATE_PRIMARY_QUERY_INDEX_TEST_XML);
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/CreatePrimaryQueryIndexChangeTest.java:41]: I think this part is ok to be in @BeforeEach method. And the same below

Comment (tigrankhojoyan) [reply liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/CreatePrimaryQueryIndexChangeTest.java:41]: We have 3 tests here(Should_generate_inverse_correctly, Should_generate_statement_correctly, Expects_confirmation_message_is_created_correctly), which don't use changelog and that step will be additional for them. I suggest to let it as is, what you think?

Comment (DaimonCool) [reply liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/CreatePrimaryQueryIndexChangeTest.java:41]: Agree. Didn't notice

**Review by DaimonCool — APPROVED** [review]

**Pull request merged**